exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/efgs/OperationDao.java

```diff
@@ -93,8 +107,10 @@ public void markErrorOperation(long operationId, Optional<String> batchTag, Time
         ), Timestamp.class);
     }
 
-    public long startOperation(EfgsOperationDirection direction) {
-        return startOperation(direction, new Timestamp(Instant.now().toEpochMilli()));
+    public Optional<Long> startOperation(EfgsOperationDirection direction, Optional<String> batchTag) {
```

Comment:
Didn't go through the use-places, but is there a need to combine these two fetches under one method? This looks like a typically multi-use function. Having a startOperation(direction) and startOperation(direction,non-optional tag) is typically cleaner and it would seem that when you're calling the method, you either have the tag or you don't -> you know what to call. 

Or is it intentional that when you have a tag and fail to find anything, you get the same result as when you don't have a tag at all? Maybe that orElseGet could be done in caller-side, if there aren't transaction things in the way.

Reply:
Should this be @Transactional, btw?

Reply:
There is only one statement on both calls, so it shouldn't be necessary? At least I think, it's not. However, there is no reason why it cannot be in there.

About api, it might be possible separate inbound and outbound, because the context is little bit different. I'll check this out later. about this tag-thingy it's fail safe for not to download everything from eu-server. Batches is an linked list and if you don't know the batch tag, then you just need to download the first batch and then you get the tag and also next tag, so I think it's cleaner in here.

Reply:
I decided to add @Transactional for all public methods, maybe it's easy to forgot it later if you add something. After second I also added it to inbound keys batch update, so it's all or nothing.

Reply:
It could also be at class level, but maybe it's good to be explicit in this case.

Reply:
Mostly I was thinking that does the first operation change something in the DB that should be rolled back if we get a disconnect in the middle of the second, etc. I don't think so, since the second is a fallback for when the first does nothing, so there should be nothing to roll back.

I wouldn't favor putting transactional to everything. I think the annotation is a good indication that multiple things happen here and the reader needs to pay attention. Most of the tasks here are so simple that there's no need for transactional, and that's always the first goal. Of course, the transactional doesn't actually hurt anything either, so it's fair to disagree.